APE9.rst

```diff
@@ -0,0 +1,273 @@
+NDArr object, and the bounding box used to obtain the slice. When involving 
+binary arithmetic operations, between two NDArr objects that have different
+parents (or are themselves distinct parents), there will be no pointer to
+the original parent(s) present, nor will the bounding box information be
+retained
+
+If operations are unary, parentage and bounding box information is propagated.
+For binary operations with two NDArr objects, operations
+will retain parentage and bounding boxes only if both operands share the same
+parents and bounding box. If they don't, this information will be removed from
+the result.
+
+The NDArr object will permit indexing relative to the parent through a special
+attribute (details at 11).
+
+NDData Details
+++++++++++++++
+
+Composition instead of Inheritance
+----------------------------------
+
+NDData will not be a subclass of NDArr. Instead it will contain references
+to an NDArr object. To avoid confusion, the attribute that contains the 
+NDArr object will be .arr
+
+Need for ExposureInfo object?
+-----------------------------
+
+This is used by LSST to store information about the exposure such as filters,
+exposure time, etc. but all that is really needed is an isomorphic mapping
+to corresponding information in the astropy object. The question is: does
+it make sense to encapsulate that information in a single attribute, and 
+whether all think that these attributes are "astropy-universal”. For the 
+purposes of this, the only standard attribute is most likely the WCS.
+
+Items such as exposure time and filters could be relegated to the meta
+attribute. Making attributes such as these at the top level could be done
+through subclasses, though it may lead to many variants. Perhaps the best way
+to deal with this is for applications or libraries to state their requirements
+for items required to be in meta and leave it at that.
+
+NDData Slicing and Related
+++++++++++++++++++++++++++
+
+Slicing of NDData is supported (eventually) as well as striding. The corresponding 
+WCS attribute, if present, is updated accordingly. Eventually there may be support
+for slicing based on world coordinates, but this will not be part of the initial
+implementation.
+
+NDData does not support mathematical operations. Utilities will be provided to aid
+in how metadata items are to be propagated when the user merges NDData instances.
+
+Mask Object
++++++++++++
+
+The mask attribute now will be an instance of a Mask Object (but to support
+backward compatibilty, the attribute must have a different name than mask since
+the current implementation treats that attribute as a boolean array (or any
+object that mimics one). [An alternative option is to assume that the mask
+attribute of NDArr is a callable whereas the mask attribute of NDData represents
+a boolean array; this would be workable, but also likely confusing]
+This APE proposes that the new attribute have the name: lone_ranger. 
+Well, perhaps dmask.
+
```

Comment:
IIRC, `FlagCollection` was, conceptually, supposed to be able to hold things like data quality masks, but its original implementation was minimal. In the [last NDData APE](https://github.com/astropy/astropy-APEs/pull/8) not a lot was said about flags or `FlagCollection`. When the code was re-written for that APE I left the `FlagCollection` code intact largely because it seemed to be doing no harm.

What is, I think, clearly missing from the NDData structure now is a consistent way to handle data quality bit planes. This comes up both for LSST and for HST (see astropy/ccdproc#389 by @SaraOgaz). 

I think the addition of that ability somewhere in the class hierarchy is useful.